src/main/host/affinity.c

```diff
@@ -0,0 +1,74 @@
+#include "affinity.h"
+
+#ifndef _GNU_SOURCE
+#define _GNU_SOURCE // Make sure this is defined for CPU_* macros
+#endif              // _GNU_SOURCE
+
+#include <assert.h>
+#include <glib.h>
+#include <sched.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <sys/sysinfo.h>
+
+#include "main/core/support/options.h"
+#include "support/logger/logger.h"
+
+// Compute once and cache this value. Apparently fetching is expensive and
+// requires parsing procfs.
+static int _ncpus = -1; // Starts uninitialized.
+
+static gint _affinity_enabled = 0;
+
+OPTION_EXPERIMENTAL_ENTRY(
+    "cpu-pin", 0, 0, G_OPTION_ARG_INT, &_affinity_enabled,
+    "Pin workers to fixed CPU cores, pin plugins to the worker's core", "[0|1]")
+
+__attribute__((constructor)) static void _init_ncpus() { _ncpus = get_nprocs(); }
+
+int affinity_getGoodWorkerAffinity(guint worker_thread_id) {
+    assert(_ncpus > 0);                    // Our host better have at least one CPU.
+    return (int)worker_thread_id % _ncpus; // Uniform distribution of workers over CPUs
+}
+
+int affinity_setProcessAffinity(pid_t pid, int new_cpu_num, int old_cpu_num) {
+    assert(_ncpus > 0 && pid >= 0);
+
+    // We can short-circuit if there's no work to do.
+    if (!_affinity_enabled || new_cpu_num == AFFINITY_UNINIT || new_cpu_num == old_cpu_num) {
+        return old_cpu_num;
+    }
+
+    cpu_set_t* cpu_set = NULL;
+    bool set_affinity_suceeded = false;
+    int retval = new_cpu_num;
+
+    if (new_cpu_num < _ncpus) { // Good, we are trying to set to a valid CPU
+        cpu_set = CPU_ALLOC(_ncpus);
+
+        if (cpu_set) { // Good, the CPU set allocation succeeded
+
+            size_t cpu_set_size = CPU_ALLOC_SIZE(_ncpus);
+
+            // Clear the CPU set
+            CPU_ZERO_S(cpu_set_size, cpu_set);
+            // And then add the new_cpu_num as the only element of the set
+            CPU_SET_S(new_cpu_num, cpu_set_size, cpu_set);
+
+            int rc = sched_setaffinity(pid, cpu_set_size, cpu_set);
+
+            set_affinity_suceeded = (rc == 0);
+        }
+    }
+
+    if (!set_affinity_suceeded) {
+        warning("Could not set CPU affinity for PID %d to %d", (int)pid, new_cpu_num);
```

Comment:
If we use the `cpu-pin` option and we can't pin, should that be an error or a warning? IIRC, Jim thought error makes more sense, rather than continuing to run the simulation without pinning and potentially tricking the user that pinning is active when it isn't. I'm thinking that logging a warning makes sense but then refusing to run the experiment. That's more work to code because you'll have to propagate the error back, but does it give us the behavior we want?

Reply:
I promoted to critical so that the user has a higher likelihood of seeing.

Crashing with an error makes me nervous because there are potentially thousands of affinity changes that could occur over the course of an experiment (during work stealing, for example). If a single one of those affinity changes doesn't succeed, I'm not sure the experiment should crash since it could otherwise complete successfully with negligible performance impact.

Regardless, if we want to make this a crash instead of a warning, I think it should be done at the call site and not inside this module.

Reply:
> Crashing with an error makes me nervous because there are potentially thousands of affinity changes that could occur over the course of an experiment

Haven't looked at the code in this PR yet but when I made that comment I believe the proposed pinning was all done during startup. If something goes wrong during startup I think it makes sense to crash. I agree we want to avoid crashing in the middle of a simulation if we can.